# 1. Data Exploration

 ## 1.1. Imports

In [8]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Set display options to show all rows
pd.set_option('display.max_rows', None)

# Set display options to show all columns
pd.set_option('display.max_columns', None)

## 1.2 Loading Data

In [9]:
data = pd.read_csv("../Data/Case1_HotelCustomerSegmentation.csv", sep=";")

In [10]:
# Display a sample of the dataset
data.sample(5)

ID Nationality   Age  DaysSinceCreation  \
32960    32961         NOR  20.0                817   
32069    32070         DEU  59.0                830   
78422    78423         PRT   NaN                351   
104323  104324         BEL  45.0                 95   
24414    24415         GBR  54.0                942   

                                                 NameHash  \
32960   0x4CF7F7F0FA91B5EB16A8D1807C8C834F2D95D1021A94...   
32069   0x59684A037D012B7F11B7EF023491B5B19C51A9052514...   
78422   0xC2C334C26C9B93A82410CA44F6CA38FA54D55CF42D79...   
104323  0x78150342F2F44C14DA45033472165079A99A65B8A1E3...   
24414   0xED9C240A5F010895B633851D48325E86E768092F6D06...   

                                                DocIDHash  AverageLeadTime  \
32960   0x82C8ACC3313BA69B6F0BB695D676E83B0FF129771322...              189   
32069   0x337132AF46188EEDEEE019F0E53067841D1CB54617F8...              237   
78422   0x9B4B6C94757081E096FCE48314EC722AEE1D64D9FB14...                0   
104323  0xD499EC6646448B7C8500B68B7F63B073D2E1177B24CA...              171   
24414   0x135811FC2AEA536A7ECFA2CB75A77EA8A44FBAFBEA29...               96   

        LodgingRevenue  OtherRevenue  BookingsCanceled  BookingsNoShowed  \
32960            295.2         164.0                 0                 0   
32069            415.8          63.0                 0                 0   
78422              0.0           0.0                 0                 0   
104323           524.4         180.0                 0                 0   
24414            201.6          68.5                 0                 0   

        BookingsCheckedIn  PersonsNights  RoomNights    DistributionChannel  \
32960                   1              8           2                 Direct   
32069                   1              9           3  Travel Agent/Operator   
78422                   0              0           0                 Direct   
104323                  1              8           4  Travel Agent/Operator   
24414                   1              6           3                 Direct   

       MarketSegment  SRHighFloor  SRLowFloor  SRAccessibleRoom  \
32960         Direct            0           0                 0   
32069          Other            0           0                 0   
78422         Direct            0           0                 0   
104323         Other            0           0                 0   
24414         Direct            0           0                 0   

        SRMediumFloor  SRBathtub  SRShower  SRCrib  SRKingSizeBed  SRTwinBed  \
32960               0          0         0       0              0          0   
32069               0          0         0       0              0          1   
78422               0          0         0       0              0          0   
104323              0          0         0       0              0          0   
24414               0          0         0       0              1          0   

        SRNearElevator  SRAwayFromElevator  SRNoAlcoholInMiniBar  SRQuietRoom  
32960                0                   0                     0            0  
32069                0                   0                     0            0  
78422                0                   0                     0            0  
104323               0                   0                     0            1  
24414                0                   0                     0            0

In [11]:
# Set ID as index
data.set_index('ID', inplace=True)

| #   | Column Name        | Description                                                                                       |
|:----|:-------------------|:--------------------------------------------------------------------------------------------------|
| 1   | ID        | Customer ID customer. |
| 2   | Nationality    | Nationality of the customer in ISO 3166-1 (Alpha 3) format. |
| 3   | Age       | Age of the customer. |
| 4   | DaysSinceCreation       | Number of elapsed days since the customer was created. |
| 5   | NameHash | Hash of the customer's name. |
| 6   | DocIDHash | Hash of the customer’s personal document identification number (usually a passport or ID card). |
| 7   | AverageLeadTime | Average number of days before arrival date the customer makes bookings. |
| 8   | LodgingRevenue | Total amount of lodging revenue paid by the customer so far. |
| 9   | OtherRevenue | Total amount of other revenue (e.g., food & beverage, spa, etc.) paid by the customer so far. |
| 10  | BookingsCanceled | Number of bookings the customer made but subsequently canceled. |
| 11  | BookingsNoShowed | Number of bookings the customer made but subsequently made a "no-show". |
| 12  | BookingsCheckedin | Number of bookings the customer made, which actually ended up staying. |
| 13  | PersonNights | Total person/nights the customer has stayed at the hotel so far. Persons/Nights are the sum of Adults and Children in each booking, multiplied by the number of Nights (Length-of-stay) of the booking. |
| 14  | RoomNights | Total of room/nights the customer has stayed at the hotel so far. Room/Nights are the multiplication of the number of rooms of each booking by the the number of Nights (Length-of- stay) of the booking. |
| 15  | DistributionChannel | Distribution channel normally used by the customer to make bookings at the hotel. |
| 16  | MarketSegment | Current market segment of the customer. |
| 17  | SRHighFloor | Indication if the customer usually asks for a room in a higher floor (0: No, 1: Yes). |
| 18  | RLowFloor  | Indication if the customer usually asks for a room in a lower floor (0: No, 1: Yes).  | 
| 19  | SRAccessibleRoom | Indication if the customer usually asks for an accessible room (0: No, 1: Yes). |
| 20  | SRMediumFloor | Indication if the customer usually asks for a room in a middle floor (0: No, 1: Yes). |
| 21  | SRBathtub | Indication if the customer usually asks for a room with a bathtub (0: No, 1: Yes). |
| 22  | SRShower | Indication if the customer usually asks for a room with a shower (0: No, 1: Yes). |
| 23  | SRCrib  | Indication if the customer usually asks for a crib (0: No, 1: Yes). |
| 24  | SRKingSizeBed | Indication if the customer usually asks for a room with a king size bed (0: No, 1: Yes). |
| 25  | SRTwinBed | Indication if the customer usually asks for a room with a twin bed (0: No, 1: Yes). |
| 26  | SRNearElevator | Indication if the customer usually asks for a room near the elevator (0: No, 1: Yes). |
| 27  | SRAwayFromElevator | Indication if the customer usually asks for a room away from the elevator (0: No, 1: Yes). |
| 28  | SRNoAlcoholInMiniBar | Indication if the customer usually asks for a room with no alcohol in the mini bar (0: No, 1: Yes). |
| 29  | SRQuietRoom | Indication if the customer usually asks for a room away from the noise (0: No, 1: Yes). |


# EDA

In [12]:
# Check the unique market segments
marketsegments = data['MarketSegment'].unique()
print(f'The {len(marketsegments)} unique market segments are: {", ".join(map(str, marketsegments))}.')

The 7 unique market segments are: Corporate, Travel Agent/Operator, Other, Direct, Complementary, Groups, Aviation.


The MarketSegment feature will be dropped, as we aim to develop a new customer segmentation model. Retaining this feature would not contribute to our analysis.

In [13]:
# Dropping the MarketSegment feature
data.drop('MarketSegment', axis=1, inplace=True)

In [14]:
# Check the number of duplicates in DocIDHash
data['DocIDHash'].duplicated().sum()
print(f'There are {data["DocIDHash"].duplicated().sum()} duplicates in the DocIDHash column.')

There are 8252 duplicates in the DocIDHash column.


In [15]:
# Drop features where DocIDHash is duplicated
data.drop_duplicates(subset='DocIDHash',
                     keep='first',
                     inplace=True) # Keep the first occurrence of the duplicated value  

After removing duplicates based on DocIDHash, we will drop both NameHash and DocIDHash, as these are solely identification features and do not contribute to the segmentation.

In [16]:
# Drop columns relative to identification (NameHash, DocIDHash)
data.drop(['NameHash', 'DocIDHash'], axis=1, inplace=True)

We will create a feature named TotalRevenue, so we can filter customers(people who have spent) from accounts registered(people who haven't spent)

In [17]:
# Create a feature called TotalRevenue which is the sum of LodgingRevenue and OtherRevenue
data['TotalRevenue'] = data['LodgingRevenue'] + data['OtherRevenue']

In [18]:
# Check how many customers have a TotalRevenue of 0 and how many have bigger than zero
nospendingdata = data[data['TotalRevenue']==0].shape[0]
print(f'There are {nospendingdata} customers with no spending data, which is {nospendingdata/data.shape[0]*100:.2f}% of the dataset.')

There are 31349 customers with no spending data, which is 30.29% of the dataset.


Approximately 30% of the dataset comprises accounts with zero spending. To focus solely on actual customers, we will filter the dataset using TotalRevenue, retaining only accounts that have made transactions at the hotel. This refined dataset will be used for all subsequent analyses.

In [19]:
# Create a dataframe with only customers who have spent
customers = data[data['TotalRevenue']>0]

In [20]:
customers.describe().T

count        mean         std   min      25%     50%  \
Age                   71914.0   48.039853   15.306212 -10.0   36.000   48.00   
DaysSinceCreation     72132.0  687.123205  375.326269  36.0  376.000  709.00   
AverageLeadTime       72132.0   89.341374   90.277544  -1.0   19.000   61.00   
LodgingRevenue        72132.0  413.594993  396.424439   0.0  196.200  317.50   
OtherRevenue          72132.0   94.212131  138.821250   0.0   28.000   60.50   
BookingsCanceled      72132.0    0.001733    0.078526   0.0    0.000    0.00   
BookingsNoShowed      72132.0    0.000416    0.022336   0.0    0.000    0.00   
BookingsCheckedIn     72132.0    1.037293    0.587137   1.0    1.000    1.00   
PersonsNights         72132.0    6.325556    4.372742   0.0    3.000    6.00   
RoomNights            72132.0    3.188294    2.088139   1.0    2.000    3.00   
SRHighFloor           72132.0    0.042367    0.201426   0.0    0.000    0.00   
SRLowFloor            72132.0    0.001234    0.035105   0.0    0.000    0.00   
SRAccessibleRoom      72132.0    0.000250    0.015795   0.0    0.000    0.00   
SRMediumFloor         72132.0    0.000665    0.025788   0.0    0.000    0.00   
SRBathtub             72132.0    0.002981    0.054514   0.0    0.000    0.00   
SRShower              72132.0    0.001733    0.041593   0.0    0.000    0.00   
SRCrib                72132.0    0.010661    0.102701   0.0    0.000    0.00   
SRKingSizeBed         72132.0    0.357054    0.479134   0.0    0.000    0.00   
SRTwinBed             72132.0    0.150100    0.357172   0.0    0.000    0.00   
SRNearElevator        72132.0    0.000347    0.018614   0.0    0.000    0.00   
SRAwayFromElevator    72132.0    0.003674    0.060501   0.0    0.000    0.00   
SRNoAlcoholInMiniBar  72132.0    0.000139    0.011774   0.0    0.000    0.00   
SRQuietRoom           72132.0    0.088685    0.284290   0.0    0.000    0.00   
TotalRevenue          72132.0  507.807124  471.830200   1.0  251.675  394.95   

                         75%       max  
Age                     59.0    115.00  
DaysSinceCreation     1005.0   1385.00  
AverageLeadTime        136.0    588.00  
LodgingRevenue         492.0  21781.00  
OtherRevenue           116.0   8859.25  
BookingsCanceled         0.0     15.00  
BookingsNoShowed         0.0      2.00  
BookingsCheckedIn        1.0     76.00  
PersonsNights            8.0    116.00  
RoomNights               4.0    185.00  
SRHighFloor              0.0      1.00  
SRLowFloor               0.0      1.00  
SRAccessibleRoom         0.0      1.00  
SRMediumFloor            0.0      1.00  
SRBathtub                0.0      1.00  
SRShower                 0.0      1.00  
SRCrib                   0.0      1.00  
SRKingSizeBed            1.0      1.00  
SRTwinBed                0.0      1.00  
SRNearElevator           0.0      1.00  
SRAwayFromElevator       0.0      1.00  
SRNoAlcoholInMiniBar     0.0      1.00  
SRQuietRoom              0.0      1.00  
TotalRevenue           606.0  23365.00

## Age
`Age:` <br>
Has missing values; <br>
Average age is 45 years; <br>
Min is -10; <br>
Max is 115.

Actionables: Deal with missing values, plot age distribution, check cases where age is negative, check cases where age is above 100. <br>


**TO DO: Create bins with age segments.**

Age wil be feature 1 of demographic clustering and there will be three groups:
- Minors: under 18;
- Young: 18-40;
- Old: <40

In [ ]:
# Check number of missing values in age
customers['Age'].isnull().sum()
print(f'There are {customers["Age"].isnull().sum()} missing values in the Age column, which correspond to {customers["Age"].isnull().sum()/len(customers)*100:.2f}% of the data.')

In [ ]:
# Check age correlation with other columns 
customers.select_dtypes(include=[np.number]).corr()['Age'].sort_values(ascending=False)

Our goal by checking this correlation was to see if any variable was correlated with age enough, that we could use in a predictive model to predict age. Since the highest correlated variable is 26% correlation, we discarded this idea, and just assumed to input median.

In [ ]:
# Check cases where age is negative
print(f'There are {len(customers[customers['Age'] < 0])} cases where age is negative')
customers[customers['Age'] < 0]

All rows with negative values for age have in common DistributionChannel = 'Travel Agent/Operator', and are between [-10, -5].


In [ ]:
# Check cases where age is above 100
print(f'There are {len(customers[customers['Age'] > 90])} cases where age is above 90 years old, which corresponds to {len(customers[customers["Age"] > 90])/len(customers)*100:.2f}% of the data.')
customers[customers['Age'] > 90]

Since only 0.02% of the data is above 90 years of age, we will treat this customers as outliers, and drop them to keep the integrity of the statistics.

In [16]:
# Drop rows where age is above 90
customers = customers[customers['Age'] <= 90]

In [17]:
# Turn the rows with age below 0 to NaN
customers.loc[customers['Age'] < 0, 'Age'] = np.nan

In [ ]:
# Fill missing values with median
customers['Age'].fillna(customers['Age'].median(), inplace=True)

In [ ]:
# check for missing values in age
print(f'There are {customers["Age"].isnull().sum()} missing values left in Age column.')

## TotalRevenue

TotalRevenue wil be feature 2 of demographic clustering and there will be three groups:
- Low Spenders: under the 33th quantile;
- Spenders: from 33th quantile to the 66th;
- Top spenders : more than the the 66th quantile

In [ ]:
# check missing values in total revenue
print(f'There are {customers["TotalRevenue"].isnull().sum()} missing values in TotalRevenue column.')

In [ ]:
# check outliers in total revenue with a horizontal boxplot
plt.figure(figsize=(10,5))
sns.boxplot(customers['TotalRevenue'], orient='h')
plt.show()

In [ ]:
# get the 99th percentile of TotalRevenue
percentile_99 = customers['TotalRevenue'].quantile(0.99)
print(f'Only 1% of customers spend more than {percentile_99:.2f}.')

In [ ]:
# How many values are above 10000

print(f'There are only {customers[customers["TotalRevenue"] > 10000].shape[0]} values above 10000 in TotalRevenue column, which represent {round(customers[customers["TotalRevenue"] > 10000].shape[0]/customers.shape[0]*100,2)}% of customers.')

We will drop customers who have spent over 10.000, as they are considered outliers.

In [24]:
# Drop rows with TotalRevenue above 10000
customers = customers[customers['TotalRevenue'] <= 10000]

In [ ]:
#Get quantile 0.33 and 0.66

quantile_33 = customers['TotalRevenue'].quantile(0.33)
quantile_66 = customers['TotalRevenue'].quantile(0.66)
print(f'The 33rd percentile is {quantile_33:.2f} and the 66th percentile is {quantile_66:.2f}.')

## BookingsCheckedIn

We will have three groups:
People who used our services (bar, spa, etc) but didn't stay at the hotel (BookingsCheckedIn)=0 <br>
One time travellers BookingsCheckedIn=1 <br>
Recurrent travellers BookingsCheckedIn>1

In [ ]:
# Check for custoemrs with BookingsCheckedIn=0
print(f'There are {customers[customers["BookingsCheckedIn"]==0].shape[0]} customers with BookingsCheckedIn=0, which is {customers[customers["BookingsCheckedIn"]==0].shape[0]/customers.shape[0]*100:.2f}% of the data.')

In [ ]:
# Check for duplicates in DocIDHash in the customers dataset that also have BookingsCheckedIn=0
print(f'There are {customers[customers["BookingsCheckedIn"]==0]["DocIDHash"].duplicated().sum()} duplicates in the DocIDHash column for customers with BookingsCheckedIn=0.')
print(f'There are {customers["DocIDHash"].duplicated().sum()} duplicates in the DocIDHash column.')

### Features for demographic clustering

Objective: Identify customer groups based on who they are (age, nationality, loyalty, and spending habits).

Feature 1 groups (Age): Young(18-35), middle aged (36-55()) or old(over 56). <br>
Feature 2 groups (TotalRevenue): Very high spender(the top 10%) , High spender(from average spent(50%) to the top 10%), budget traveler (below average spent (lowest 50%)). <br>
Feature 3 groups (BookingsCheckedIn): one time stop(=1), infrequent(1-3), frequent (<3)  <br>
Feature 4 groups (PeoplePerRoom = PersonsNights/RoomNights): solo traveler(1), small group(2-5), big group (<5).